qui/updater/progress_page.py

```diff
@@ -157,14 +156,36 @@ def update_admin_vm(self, admins):
 
         try:
             with Ticker(admin):
-                untrusted_output = subprocess.check_output(
-                    ['sudo', 'qubesctl', '--dom0-only', '--no-color',
-                     'pkg.upgrade', 'refresh=True'],
-                    stderr=subprocess.STDOUT)
-                output = self._sanitize_line(untrusted_output)
-
-                GLib.idle_add(admin.append_text_view, output)
-                GLib.idle_add(admin.set_status, UpdateStatus.Success)
+                # pylint: disable=consider-using-with
+                proc = subprocess.Popen(
+                    ['sudo', 'qubes-dom0-update'],
+                    stderr=subprocess.PIPE, stdout=subprocess.PIPE)
+
+                read_err_thread = threading.Thread(
+                    target=self.dump_to_textview,
+                    args=(proc.stderr, admin)
+                )
+                read_out_thread = threading.Thread(
+                    target=self.dump_to_textview,
+                    args=(proc.stdout, admin)
+                )
+                read_err_thread.start()
+                read_out_thread.start()
+
+                while proc.poll() is None \
+                        or read_out_thread.is_alive() \
+                        or read_err_thread.is_alive():
+                    time.sleep(1)
+                    if self.exit_triggered and proc.poll() is None:
+                        proc.send_signal(signal.SIGINT)
+                        proc.wait()
+                        read_err_thread.join()
+                        read_out_thread.join()
+
+                if "No updates available" in admin.buffer:
```

Comment:
This works only together with `--check-only` option. Without `--check-only`, I see `Nothing to do.` message, but I don't think it's very reliable (it may also show up if some updates already already downloaded in updatevm, but not installed yet, for example). `--check-only` + exit code checking should be more reliable here than running full update and parsing the outpu, but requires two calls:
1. `sudo qubes-dom0-update --refresh --check-only` and check exit code:
  - 0: no updates
  - 100: updates available
  - anything else: error
2. `sudo qubes-dom0-update -y` (without --refresh this time) to apply updates

Alternatively, maybe `qubes-dom0-update` can be extended to do that internally, like have an option to signal "some updates successfully installed" via specific exit code (may be also 100?).